### DISCUSSION TITLE
Trying to use pure mode with a header file

### DISCUSSION THREAD

**Discussion by mike919192** [hsutter/cppfront#1066] [Q&A] [answered]
Hi.  I've had success with cppfront in mixed mode, but decided to try a simple example in pure mode and I'm running into an issue.  I can use pure mode with a single cpp2 file, but I am getting an error trying to include a h2 file.

hello.cpp2
```
#include "hello.h2"

main: () = {
    
    std::cout << "Hello " << name() << "\n";
}
```
hello.h2
```
name: () -> std::string = {
    s: std::string = "world";
    decorate(s);
    return s;
}

decorate: (inout s: std::string) = {
    s = "[" + s + "]";
}
```
Compiler output:
```
abc@4e0512100c8d:~/workspace/test_cppfront$ ../cppfront/source/cppfront -p *.cpp2 *.h2
hello.cpp2... ok (mixed Cpp1/Cpp2, Cpp2 code passes safety checks)

hello.h2... ok (all Cpp2, passes safety checks)

abc@4e0512100c8d:~/workspace/test_cppfront$ g++ -I../cppfront/source -std=c++20 -g *.cpp -o hello
/usr/bin/ld: /tmp/ccVKnHmH.o: in function `main':
/config/workspace/test_cppfront/hello.cpp2:8: undefined reference to `name[abi:cxx11]()'
collect2: error: ld returned 1 exit status
abc@4e0512100c8d:~/workspace/test_cppfront$ 
```
I also noticed that the h2 file is generating both a h file and also hpp file.  

Am I doing something wrong trying to include the header?  If I move the function definitions into the cpp2 then the pure mode is working.

**Comment by vovaskochko** [comment]
Have you tried to include file name of generated header instead of #include "hello.h2" ?

**Reply by mike919192** [reply]
I could be wrong but I don't think that's the intended usage.  When pure mode is on you get this message if you include anything except for h2 file.

`hello.h2(1,1): error: pure-cpp2 switch disables the preprocessor, including #include (except of .h2 files) - use import instead (note: 'import std;' is implicit in -pure-cpp2)`

**Reply by vovaskochko** [reply, marked as answer]
Some time ago I tried to use h2 headers and I faced with the issue. In my case I just opened the generated file and observed that include is still h2. But I did not use the pure mode that time as I remember.
Also I didnt find documentation on this so I performed some checks of source code that time

**Reply by mike919192** [reply]
Ah OK, I think you are on the right track.  The generated cpp file includes h but does not include the hpp.  This does seem to be a bug then because I add the include statement to the generated cpp and rerun g++ and it is working.

hello.cpp
```
#define CPP2_IMPORT_STD          Yes

//=== Cpp2 type declarations ====================================================


#include "cpp2util.h"

#line 1 "hello.cpp2"


//=== Cpp2 type definitions and function declarations ===========================

#line 1 "hello.cpp2"
//#include <iostream>
//#include <string>

#include "hello.h"
#include "hello.hpp" //this was missing

#line 6 "hello.cpp2"
auto main() -> int;

//=== Cpp2 function definitions =================================================

#line 1 "hello.cpp2"

#line 6 "hello.cpp2"
auto main() -> int{

    std::cout << "Hello " << name() << "\n";
}
```

**Comment by GrahamAsher** [comment]
I have fixed the problem in my fork of cppfront by moving the function which emits `#include "myheader.hpp" ` out of a conditional block where I don't think it belongs. Please note that I haven't analysed the code, and although it works for me, it may not be the correct fix. But, for what it is worth, here is the patch:

```
 source/to_cpp1.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/source/to_cpp1.h b/source/to_cpp1.h
index 75880d5..b7f713d 100644
--- a/source/to_cpp1.h
+++ b/source/to_cpp1.h
@@ -1519,9 +1519,9 @@ public:
             printer.print_extra( "\n#ifndef " + cpp1_FILENAME+"_CPP2" );
             printer.print_extra( "\n#define " + cpp1_FILENAME+"_CPP2" + "\n\n" );
 
-            printer.print_extra( hpp_includes );
         }
 
+        printer.print_extra(hpp_includes);
 
         //---------------------------------------------------------------------
         //  Do phase2_func_defs
```
